src/pubsub.js

```diff
@@ -50,6 +61,65 @@ class BasicPubSub extends Pubsub {
     })
   }
 
+  /**
+   * Dial a received peer.
+   * @override
+   * @param {PeerInfo} peerInfo peer info
+   * @param {function} callback
+   *
+   * @returns {void}
+   */
+  _dialPeer (peerInfo, callback) {
+    callback = callback || function noop () { }
+    const idB58Str = peerInfo.id.toB58String()
+
+    // If already have a PubSub conn, ignore
+    const peer = this.peers.get(idB58Str)
+    if (peer && peer.isConnected) {
+      return nextTick(() => callback())
+    }
+
+    // If already dialing this peer, ignore
+    if (this._dials.has(idB58Str)) {
+      this.log('already dialing %s, ignoring dial attempt', idB58Str)
+      return nextTick(() => callback())
+    }
+    this._dials.add(idB58Str)
+
+    this.log('dialing %s', idB58Str)
+
+    // Define multicodec to use
+    let multicodec = this.multicodec
+
+    // Should fallback to floodsub if no Gossipsub and fallback enabled
+    if (this._options.fallbackToFloodsub &&
+      !peerInfo.protocols.has(this.multicodec)) {
```

Comment:
The issue here is that if we are dialing the peer for the first time we won't know their protocols, so this is going to return false and use floodsub even if gossipsub is actually supported.

I think this logic can be moved inside of `dialProtocol`. The idea being that we can attempt to negotiate gossipsub right away. If the dial fails and we get an `err`, we should be able to see if this is due to a protocol unsupported error. Since the `Identify` protocol runs in parallel, we're not guaranteed to know the protocols yet, so we should inspect the error (we will probably want to update the multistream-select error to provide a code as I dont believe it currently does).

Additionally, if we wanted to try and avoid blind protocol negotiation when possible, we could check to see if the `peerInfo.protocol` set is empty, and if not check the protocols for our desired list of pubsub protocols (gossip, then flood).

Reply:
Thanks for the suggestions Jacob!
Mind having a look at the new approach? I think that it encompasses all your thoughts 🙂